### ISSUE TITLE
Generalising arithmetic operators in non-inline functions

### ISSUE THREAD

**Issue by Happypig375** [fsharp/fslang-suggestions#1032]
# Generalising arithmetic operators in non-inline functions

With https://github.com/dotnet/designs/pull/205, .NET is planning to expose a set of interfaces that expose arithmetic operators as well as related constants (additive identity, multiplicative identity, min/max values, etc.)

On platforms where these are defined like .NET6+ as planned, F# should be able to use them to automatically generalise instead of defaulting to `int`.

The existing way of approaching this problem in F# is being forced to use `inline` which doesn't interoperate well with C#.

## Pros and Cons 

The advantages of making this adjustment to F# are
1. Convenience
2. Consistency with the rest of automatic generalisation
3. Correctness

The disadvantages of making this adjustment to F# are that this may be a breaking change. However, who would rely on these defaults heavily?

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): M to L

Related suggestions: (none)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [ ] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by charlesroddie** [comment]
Static members in interfaces could result in some significant changes to arithmetic.

It would be very useful to allow `a + b` where `a` and `b` are `IAddable<'t>'.

I think the title of the issue represents something much more minor in importance, and while it would fix a wart in F# (incorrect type inference of `int`) it's hard to fix for backwards compat and the ability to use operators is much more important than type inference. I would leave this as a follow-on discussion once this is implemented, which could well be 1-2years since dotnet6 needs to come out and then F# needs to move to dotnet6.

Do we need a way to shelve requests which may be good ideas but cannot currently be implemented? Close and then reopen when implementable?

**Comment by cartermp** [comment]
We're going to eventually support this in some form, for interop, when the feature stabilizes beyond .NET 6. The .NET 6 / C# 10 feature is kind of a "phase 1" implementation that's going to span at least one more .NET release.

**Comment by charlesroddie** [comment]
There is some discussion of the application of static interface methods to generic maths here: https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/

It's worth getting input from heavy .Net maths users as the feature is in preview and can therefore get last-minute changes before .Net 6 release. The blog post there suggests where to put feedback and there is discussion in the comments. Pinging some maths library mantainers @Happypig375 @foggyfinder @whiteblackgoose @ForNeVeR @gsomix @cdrnet . As I put in the comments, I think static interface methods for maths is one of the best things I have seen in .Net, but the `INumber` interface is one of the worst things I have seen in .Net.

**Comment by ForNeVeR** [comment]
Well, considering this is still in preview in C# (and could require future changes in both language and CLR to work), I would say that it's unwise to tightly bind F# language design to this feature _at this moment_.

In the future, I think the design of this could go in the following direction:

1. Think about integration of these new static interface methods into the language (and I think these shouldn't only be used for interop; they seem to be very useful for generic code where we would use `inline` otherwise, pros being clean C# interop and cleaner F# code without any of the SRTP oddities).
2. Declare a new generic operator `(+)` requiring `IAddition<T>` somewhere in the FSharp.Core (in a module not open by default).
3. Now, a program using this operator could look like this:
   ```fsharp
   open GenericMath
   let add a b = a + b // uses the `GenericMath.(+)` operator,
                       // so have a signature of `TAddable -> TAddable -> TAddable`
   ```
   This could then work in a similar way to `open Checked` (also being an opt-in, at least for the beginning).

**Comment by dsyme** [comment]
The IWSAM support has been merged in F# 7.0 and this sort of programming is now more feasible without inline
